akka-stream/src/main/scala/akka/stream/scaladsl/Restart.scala

```diff
@@ -37,10 +37,11 @@ object RestartSource {
    * @param randomFactor after calculation of the exponential back-off an additional
    *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
    *   In order to skip this additional delay pass in `0`.
+   * @param maxRestarts the amount of restarts is capped to this amount
    * @param sourceFactory A factory for producing the [[Source]] to wrap.
    */
-  def withBackoff[T](minBackoff: FiniteDuration, maxBackoff: FiniteDuration, randomFactor: Double)(sourceFactory: () ⇒ Source[T, _]): Source[T, NotUsed] = {
-    Source.fromGraph(new RestartWithBackoffSource(sourceFactory, minBackoff, maxBackoff, randomFactor, onlyOnFailures = false))
+  def withBackoff[T](minBackoff: FiniteDuration, maxBackoff: FiniteDuration, randomFactor: Double, maxRestarts: Int)(sourceFactory: () ⇒ Source[T, _]): Source[T, NotUsed] = {
```

Comment:
Having `maxRestarts: Int` would require to cap the amount of restarts always. I am not sure if that's a good idea. Maybe it's better to have it as an `Option[Int]` where `None` would mean that amount of restarts are not capped.

Anyway this would be a breaking change to the API, or? Should there be a default parameter (like `None`)?

Depending on what we discuss, how should the java api look like?

@hseeberger what do you think?

Reply:
So far Akka is rather using "magic numbers" (-1, Int.MaxValue, etc.) instead of `Option`s. Which is probabaly due to the lack of `Optional` prior to Java 8. Hence I'm fine with an `Int` instead of an `Option[Int]`, because it's consistent(ly bad).

Reply:
Okay, I will add some logic that no limiting is applied when `maxRestarts <= 0` (since having a RestartSource with 0 restarts does not make sense at all).

What about my other question:
Should I add a default value, e.g. `maxRestarts: Int = -1` to not break existing usages?